src/query.rs

```diff
@@ -52,4 +52,50 @@ impl Query {
             inner: Box::new(inner),
         })
     }
+
+    /// Construct a Tantivy's PhraseQuery
+    #[staticmethod]
+    #[pyo3(signature = (schema, field_name, words))]
+    pub(crate) fn phrase_query(
+        schema: &Schema,
+        field_name: &str,
+        words: &PyList,
+    ) -> PyResult<Query> {
+        let mut terms = Vec::<tv::Term>::new();
+        for w in words.iter() {
+            let term = make_term(&schema.inner, field_name, w)?;
+            terms.push(term);
+        }
+        let inner = tv::query::PhraseQuery::new(terms);
+        Ok(Query {
+            inner: Box::new(inner),
+        })
+    }
+
+    /// Construct a Tantivy's PhraseQuery with custom offsets and slop
+    #[staticmethod]
+    #[pyo3(signature = (schema, field_name, words, offsets, slop = 0))]
+    pub(crate) fn phrase_query_offset_slop(
+        schema: &Schema,
+        field_name: &str,
+        words: &PyList,
+        offsets: &PyList,
+        slop: u32,
+    ) -> PyResult<Query> {
+        assert!(
```

Comment:
Since we return a result anyway, I think we should make this either a just an error instead of a panic (which is rather hard to catch in Python).

Better yet, I would prefer to have this as a single function using the flexibility of Python's dynamic types, e.g.

```rust
/// Construct a Tantivy's PhraseQuery
#[staticmethod]
#[pyo3(signature = (schema, field_name, words, slop = 0))]
pub(crate) fn phrase_query(
    schema: &Schema,
    field_name: &str,
    words: &PyAny,
    slop: u32,
) -> PyResult<Query> {
    let mut terms_with_offset = Vec::<(usize, tv::Term)>::new();
    let mut terms = Vec::<tv::Term>::new();
    for word in words.iter()? {
        if let Ok((offset, word)) = word.extract() {
            let term = make_term(&schema.inner, field_name, w)?;
            terms_with_offset.push((offset, term));
        } else {
            let term = make_term(&schema.inner, field_name, w)?;
            terms.push(term);
        }
    }
    let inner = if !terms_with_offset.is_empty() && terms.is_empty() {
        tv::query::PhraseQuery::new_with_offset(terms_with_offset)
    } else if terms_with_offset.is_empty() && !terms.is_empty() {
        tv::query::PhraseQuery::new(terms)
    } else {
        return Err(PyValueErr::new_err("Provide either offsets for all terms or for none"));
    }
    Ok(Query {
        inner: Box::new(inner),
    })
}
```

(Have neither compiled nor tested this.)

Reply:
@adamreichold Thank you for your suggestion.

> Since we return a result anyway, I think we should make this either a just an error instead of a panic (which is rather hard to catch in Python).

I think you are totally right - sorry, I have limited experience with actual Rust projects.

Reply:
@mocobeta It is ok - we are learning together 😄

Reply:
I refined the method signature so that `words` takes either a list of str or a list of tuple[int, str].